### ISSUE TITLE
Spread operator with wrong optional property raises error on incorrect source line

### ISSUE THREAD

**Issue by laverdet** [microsoft/TypeScript#51376]
# Bug Report

### 🔎 Search Terms
spread error wrong line

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, [not applied: and I reviewed the FAQ for entries about]. I tried on v3.3.3 and also on nightly.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.9.0-dev.20221025#code/C4TwDgpgBAksEFsoF4oG8CwAoKuoENhgAnASwCMBXeALigGcTSA7AcwG5sBfTrbAYwD2zRlGIR6lADbA6ABXzFgpfFIA8AJQhDiAEzUByQkyrwDAGgZM2UAD5Rm0qQD5nKdNjwEiZUxDoARAHm3NgCwqKk8Ah0cIgomDh4xr7U-lBBIUm4AHR54pIyWVxAA)

### 💻 Code

```ts
type Item = {
    attribute: string;
};

const result: Partial<Record<'attribute', string | null>> = {
    attribute: "",
}

const item: Item ={
    attribute: "",
    ...result,
}
```

### 🙁 Actual behavior
This code is, of course, an error. The surprising thing is *where* the error is raised. In this case the line reading `attribute: "",` will be highlighted with `Type 'string | null' is not assignable to type 'string'.`

In this simple reproduction the error is clear, but when I discovered it the types were more complicated and there were a dozen or more attributes in between the scary red line and the offending spread.

### 🙂 Expected behavior
The diagnostic error would be much more helpful if the spread containing the invalid property was highlighted instead of the property which actually has correct types.

**Comment by isaacudofia** [comment]
Probably no solution until they fix it, except replacing the … operator with Object.assign

**Comment by RyanCavanaugh** [comment]
I can't reproduce the reported error location with the linked code. With `--strict`, TypeScript 3.3.3, 4.9.0-dev.20221025, 6.0.3, and the current native build all report TS2322 on `...result`, not on the preceding `attribute: ""` property.

Please provide the `tsconfig.json` settings and the editor/TypeScript extension version from the case that highlights `attribute: ""`, along with the exact source if it differs from the Playground example.

**Comment by laverdet** [comment]
Are we looking at the same playground link? The repro is still valid on v6 and v7, with and without `--strict`.

This is the output from v7:

```
marcel[1:19:45AM] [~/tmp] npx tsc --version
Version 7.0.2

marcel[1:19:49AM] [~/tmp] cat a.ts 
type Item = {
    attribute: string;
};

const result: Partial<Record<'attribute', string | null>> = {
    attribute: "",
}

const item: Item ={
    attribute: "",
    ...result,
}

marcel[1:19:52AM] [~/tmp] cat tsconfig.json        
{
  "compilerOptions": {
    "strict": true,
  }
}

marcel[1:19:57AM] [~/tmp] npx tsc -b       
a.ts:10:5 - error TS2322: Type 'string | null' is not assignable to type 'string'.
  Type 'null' is not assignable to type 'string'.

10     attribute: "",
       ~~~~~~~~~
  a.ts:2:5 - The expected type comes from property 'attribute' which is declared here on type 'Item'
    2     attribute: string;
          ~~~~~~~~~

Found 1 error in a.ts:10
```

To be clear, we expect an error in the code. The error of `Type 'string | null' is not assignable to type 'string'.` is very confusing given the highlighted source text fragment `attribute: ""`. In more complicated type aliases it can become a real head-scratcher.

**Comment by RyanCavanaugh** [comment]
Probably a clearer repro is
```ts
type Item = {
    attribute: string;
};

const result = {
    attribute: 0,
}

const item: Item = {
    attribute: "",
    ...result,
}
```
where the error span is still on the property instead of the spread. We need to compute whether the spread possibly contributed to the relevant property's type and, if it did, probably just give up and highlight the entire literal, since we don't have the ability to issue one error with two error spans.
